### ISSUE TITLE
Single visible menu is made up of multiple menus

### ISSUE THREAD

**Issue by karlgroves** [WordPress/gutenberg#15318]
Single visible menu is made up of multiple menus

-   **Severity:** High
-   **Affected Populations:**
    -   Blind
    -   Low-Vision
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Tools and Options

#### Issue description

The Tools and Options menu is split into several sub-groups, each of
which uses the ARIA menu pattern, and when a user navigates using the
Arrow keys, they are unable to proceed past each group without using Tab
(i.e. Arrow keys simple cycle through the items in one sub-group). The
user is forced to switch between Tab and Arrow to navigate through the
whole menu.

Keyboard users, especially those with cognitive disabilities, may find
the current menu unpredictable and inefficient to use, because of its
unconventional keystrokes. If they are not using a screen reader, they
will not realise that there are multiple separate menus in a single
dropdown.

Semantic markup is vital for assistive technology users, to convey the
structure and relationships between information. When appropriate
semantic markup is not used, it may be more difficult for such users to
understand what the content means, and to comprehend the structure that
the visual presentation conveys.

[![Screenshot of the content with the issue described
here](/images/17599_multiple_menu.png)]{.image-wrap}

##### Issue Code

```
    <div tabindex="-1" class="components-popover__content" style="max-height: 247.5px;">


        <div class="...">


            <div class="..." id="components-menu-group-label-0">View</div>


            <div role="menu" aria-labelledby="components-menu-group-label-0" aria-orientation="vertical">


                <button class="..." role="menuitemcheckbox" aria-checked="true" aria-describedby="edit-post-feature-toggle__info-0" aria-label="Top Toolbar" type="button">


                    ...


                </button>


                <button class="..." role="menuitemcheckbox" aria-checked="false" aria-describedby="edit-post-feature-toggle__info-1" aria-label="Spotlight Mode" type="button">


                    ...


                </button>


                <button class="..." role="menuitemcheckbox" aria-checked="false" aria-describedby="edit-post-feature-toggle__info-2" aria-label="Fullscreen Mode" type="button">


                    ...


                </button>


            </div>


        </div>


        <div class="...">


            <div class="..." id="...">Editor</div>


            <div role="menu" aria-labelledby="...." aria-orientation="vertical">


                <button class="..." role="menuitemradio" aria-checked="true" aria-label="Visual Editor" type="button">


                    ...Visual Editor


                </button>


                <button class="..." role="menuitemradio" aria-checked="false" aria-label="Code Editor" type="button">


                    Code Editor...


                </button>


            </div>


        </div>


        <div class="...">


            <div class="..." id="...">Tools</div>


            <div role="menu" aria-labelledby="..." aria-orientation="vertical">


                ...


            </div>


        </div>


        <div class="...">


            <div role="menu" aria-labelledby="..." aria-orientation="vertical">


                <button class="..." role="menuitem" aria-label="Options" type="button">Options</button>


            </div>


        </div>


    </div>
```

#### Remediation Guidance

Using a single ARIA menu
(<https://www.w3.org/TR/wai-aria-practices-1.1/#menu>) would be better
than multiple ARIA menus, because then it would have consistent keyboard
behaviours that are recognisable to both sighted keyboard users and
assistive technology users. However this pattern does not have any scope
for providing sub-groups and group headings, and the current solution of
splitting the menu is potentially confusing for users.

So it may be best to have the "Tools and Options" button open a popup,
rather than a menu, which contains a form with typical form controls in
fieldset groups. This approach is shown in the Recommended Code.

Other possible alternatives to this recommendation are:

* Single unordered-list, where each section is a list-item containing a
heading and a child list (navigated via the Tab key, not Arrow keys).
* Popup containing multiple unordered lists grouped with headings.
* Remove the group headings and implement the menu as a single ARIA
menu.

In any case, the Options interface should be navigable as a single unit,
without needing to switch between Tab and Arrow to navigate through the
whole thing.

##### Recommended Code

```
    <div tabindex="-1" class="components-popover__content" style="max-height: 247.5px;">


        <fieldset class="...">


            <legend class="..." id="components-menu-group-label-0">View</legend>


                <div>


                    <input type="checkbox" id="edit-post-feature-toggle-0" aria-describedby="edit-post-feature-toggle__info-0">


                    <label for="edit-post-feature-toggle-0">Top Toolbar"</label>


                    <span id="edit-post-feature-toggle__info-0" class="...">Access all block and document tools in a single place</span>


                </div>


                <div>


                    <input type="checkbox" id="edit-post-feature-toggle-1" aria-describedby="edit-post-feature-toggle__info-1">


                    <label for="edit-post-feature-toggle-1">Spotlight Mode</label>


                    <span id="edit-post-feature-toggle__info-1" class="...">Focus on one block at a time</span>


                </div>


                <div>


                    <input type="checkbox" id="edit-post-feature-toggle-2" aria-describedby="edit-post-feature-toggle__info-2">


                    <label for="edit-post-feature-toggle-2">Fullscreen Mode</label>


                    <span id="edit-post-feature-toggle__info-2" class="...">Work without distraction</span>


            </div>


        </fieldset>


        <fieldset class="...">


            <legend class="..." id="...">Editor</legend>


            <div>


                <input type="radio" checked id="editor_type">


                <label for="editor_type">Visual Editor</label>


            </div>


            <div>


                <input type="radio" id="editor_type-2" aria-labelledby="components-shortcut-2">


                <label for="editor_type-2">Code Editor</label>


                 <span id="components-shortcut-2">Ctrl+Shift+Alt+M</span>


            </div>


        </fieldset>


        <fieldset class="...">


            <legend class="..." id="...">Tools</legend>


            <div>


                <a href="edit.php?post_type=wp_block" class="...">Manage All Reusable Blocks</a>


                <button type="button" class="....">Keyboard Shortcuts <span class="components-menu-item__shortcut">Shift+Alt+H</span></button>


                <button type="button" class="...">Copy All Content</button>


            </div>


        </fieldset>


        <fieldsetclass="...">


            <button class="..." type="button">Options</button>


        </fieldset>


    </div>
```

#### References

<https://github.com/WordPress/gutenberg/issues/12505>

#### Relevant standards

-   [1.3.1 Info and Relationships
    (Level A)](https://www.w3.org/TR/WCAG20/#content-structure-separation-programmatic)
-   [2.1.1 Keyboard
    (Level A)](https://www.w3.org/TR/WCAG20/#keyboard-operation-keyboard-operable)
-   [4.1.2 Name, Role, Value
    (Level A)](https://www.w3.org/TR/WCAG20/#ensure-compat-rsv)

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-70 in Tenon's report

**Comment by afercia** [comment]
This is under the works and should (hopefully) land soon, see #12505,  #13059, and #14843. /Cc @aduth @gziolo

**Comment by melchoyce** [comment]
Screenshot from the full report:

![image](https://user-images.githubusercontent.com/2846578/57182840-3ca37200-6e59-11e9-922a-6241ea21e4fd.png)
